### ISSUE TITLE
Constraints-driven P_R?!

### ISSUE THREAD

**Issue by chili-chips-ba** [chili-chips-ba/openCologne#38]
GateMate P_R is apparently using  _quadratic placement_ algorithm.
- Where can we find more about GateMate proprietary `constraints-driven placement` and `constraints-driven routing`?
- What exact constraints are tapped into to drive that process?
- Any CCF examples you can share that make use of them?

> ![GateMate](https://github.com/user-attachments/assets/4650b4ee-2b11-437d-a703-954e337a03f3)
Also [see](https://github.com/chili-chips-ba/openCologne/issues/18#issuecomment-2344233355)

**Comment by chili-chips-ba** [comment]
@mmicko, any references you can share (doc or example) on this topic for the new opensource PnR flow?!

**Comment by mmicko** [comment]
Regarding documentation for nextpnr all is part od docs, and papers explaining algorithms are not ours but available and not hard to find. For code examples https://github.com/YosysHQ/prjpeppercorn-test-cases is best reference.

**Comment by chili-chips-ba** [comment]
How about specifying timing constraints, such as clock period, false paths, multi-cycle paths, async clock groups, I/O delays, etc.?

**Comment by chili-chips-ba** [comment]
Yes, it is a good idea to have that attribute for GateMate too! And to also positively report the I/O flops that, for whatever reason, could not be placed in the I/O.

**Comment by mrcmry** [comment]
It seems from the gatemate datasheet  that the ccf file has the FF_IBF and FF_OBF pin optional constraints.

With it the packer move the flip-flop into the IO bel. It seems to be implemented here:

https://github.com/YosysHQ/nextpnr/blob/2894e539343b83f27055845ee453c03dd3acdbc4/himbaechel/uarch/gatemate/pack_io.cc#L518

I have not tested it though. @Ravenslofty perhaps you can confirm it works as intended.
